src/cim/components/Agent/AgentsSelectionUtils.ts

```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import { useFormikContext } from 'formik';
+import * as _ from 'lodash';
+import { AgentK8sResource } from '../../types/k8s/agent';
+import {
+  ClusterDeploymentHostsSelectionValues,
+  ScaleUpFormValues,
+} from '../ClusterDeployment/types';
+import { AGENT_LOCATION_LABEL_KEY, AGENT_NOLOCATION_VALUE } from '../common/constants';
+
+type AllowedFormValues = ClusterDeploymentHostsSelectionValues | ScaleUpFormValues;
+
+export const useAgentsAutoSelection = <FormValues extends AllowedFormValues>(
+  availableAgents: AgentK8sResource[],
+) => {
+  const { setFieldValue, values } = useFormikContext<FormValues>();
+  const { hostCount, locations, autoSelectedHostIds } = values;
+
+  const [matchingAgents, selectedAgents] = React.useMemo(() => {
+    const mAgents = availableAgents.filter((agent) => {
+      const agentLocation =
+        agent.metadata?.labels?.[AGENT_LOCATION_LABEL_KEY] || AGENT_NOLOCATION_VALUE;
+      return locations.length ? locations.includes(agentLocation) : true;
+    });
+    const sAgents = mAgents.filter((a) => autoSelectedHostIds.includes(a.metadata?.uid || ''));
+    return [mAgents, sAgents];
+  }, [availableAgents, locations, autoSelectedHostIds]);
+
+  React.useEffect(() => {
+    const ids = matchingAgents.map((a) => a.metadata?.uid).splice(0, hostCount);
+    if (!_.isEqual(ids, autoSelectedHostIds)) {
+      setTimeout(() => setFieldValue('autoSelectedHostIds', ids, true));
+    }
+  }, [matchingAgents, setFieldValue, autoSelectedHostIds, hostCount]);
+
+  return { matchingAgents, selectedAgents, hostCount };
```

Comment:
lets return array to stay consistent with other hooks

Reply:
Hmm, returning an array forces me to explicitly specify the return value of `useAgentsAutoSelection` as `: [AgentK8sResource[], AgentK8sResource[], number]` without it typescript infers the return value as `(AgentK8sResource[] | number)[]` which does not work with the usage. Also in this case it is not expected the returned keys to need to be renamed by the consuming application so I feel using an object is better in this case.
Formik uses objects as well btw.